### ISSUE TITLE
About 25% of our jobs on Trusty error with mysql connect problems.

### ISSUE THREAD

**Issue by nerrad** [travis-ci/travis-ci#8137]
Here's some example jobs:

- https://travis-ci.org/eventespresso/ee4-events-calendar/jobs/256034551
- https://travis-ci.org/eventespresso/eea-wait-lists/jobs/256034055
- https://travis-ci.org/eventespresso/eea-barcode-scanner/jobs/255913934

The error message in the log:

```
+mysqladmin create wordpress_test --user=root --password= --host=localhost --protocol=tcp
Warning: Using a password on the command line interface can be insecure.
mysqladmin: connect to server at 'localhost' failed
error: 'Can't connect to MySQL server on 'localhost' (111)'
Check that mysqld is running on localhost and that the port is 3306.
You can check this by doing 'telnet localhost 3306'
```

Restarting the job usually fixes.  This happens across various php versions.  The only seemingly consistent pattern across all our builds that do this is running on the new trusty box with sudo:false.

**Comment by BanzaiMan** [comment]
Do the results vary within a single repository?

**Comment by nerrad** [comment]
Yes, on one build all jobs could pass successfully, on another build (for the same repo) one or two fail because of the mysql connect issue.  Also the results vary across jobs in the same repository. In other words jobs that fail with this issue in one build, pass fine in another build.

**Comment by nerrad** [comment]
Here's an example:
- https://travis-ci.org/eventespresso/event-espresso-core/builds/256110134
- https://travis-ci.org/eventespresso/event-espresso-core/builds/256053264
- https://travis-ci.org/eventespresso/event-espresso-core/builds/255913993

**Comment by SuperTux88** [comment]
I have the same problem with mysql in trusty containers. Here is an example:

* https://travis-ci.org/diaspora/diaspora/jobs/256707749#L824

`#<Mysql2::Error: Can't connect to local MySQL server through socket '/var/run/mysqld/mysqld.sock' (2)>`

When I click "Restart Job" it works the second time (I did that with some other jobs where mysql failed, that's why I have only one example).

This issue is probably a duplicate of #7790.

**Comment by nazar-pc** [comment]
I've just switched to MariaDB, even latest stable 10.2 is available. MariaDB seems to work fine.

**Comment by nazar-pc** [comment]
I've installed MariaDB 10.1 and don't have such issues (10.2 fails to install and I had different issues because of this).

**Comment by EnTeQuAk** [comment]
This is failing for us too, example jobs that failed:

 * https://travis-ci.org/mozilla/addons-server/builds/256644239
 * https://travis-ci.org/mozilla/addons-server/builds/256534665
 * https://travis-ci.org/mozilla/addons-server/builds/256823063
 * https://travis-ci.org/mozilla/addons-server/builds/256534665

We have to switch back to precise for now unfortunately because of this.

**Comment by nazar-pc** [comment]
Why MariaDB 10.1 is not a better future-proof workaround for you?

**Comment by EnTeQuAk** [comment]
> Why MariaDB 10.1 is not a better future-proof workaround for you?

If you mean me, well, we do have to test with whatever we're using in our production environment which means it's more MySQL 5.6 (we're using 5.7 but that's not easily installable in the container environments) than MariaDB.

**Comment by arvenil** [comment]
I've reported this 2 months ago: https://github.com/travis-ci/travis-ci/issues/7790
> I've opened this issue, because to be honest it looked really weird to me that MySQL for container solution fails so often... I know it's beta, but still this happens way to often. It happens so often that it made me think it's somehow fault on my side. Maybe it would be worth to monitor this on travis side to give you an idea what's the scale of the issue.

A best solution ever was to switch to docker containers:


```yaml
# .travis.yaml with matrix of MySQLs (https://docs.travis-ci.com/user/customizing-the-build/#Build-Matrix)
dist: trusty
sudo: required

language: go

go:
  - 1.8.x

services:
  - docker

env:
  global:
    - MYSQL_USER="root"
    - MYSQL_HOST="127.0.0.1"
    - MYSQL_PORT="3406"
  matrix:
    - MYSQL_IMAGE=mysql:5.5
    - MYSQL_IMAGE=mysql:5.6
    - MYSQL_IMAGE=mysql:5.7

before_script:
  # run docker containers
  - docker-compose up -d
  # wait for MySQL to become available
  - |
    until mysql -u"${MYSQL_USER}" -h "${MYSQL_HOST}" -P "${MYSQL_PORT}" -e 'SELECT 1' > /dev/null; do
      >&2 echo "MySQL ${MYSQL_HOST}:${MYSQL_PORT} for user ${MYSQL_USER} is unavailable - sleeping"
      sleep 1
    done
    >&2 echo "MySQL ${MYSQL_HOST}:${MYSQL_PORT} is up"
```

```yaml
# docker-compose.yml
version: '3'
services:
  mysql:
    image: ${MYSQL_IMAGE:-mysql}
    ports:
      - ${MYSQL_HOST:-127.0.0.1}:${MYSQL_PORT:-3306}:3306
    environment:
      - MYSQL_ALLOW_EMPTY_PASSWORD=yes
```

I didn't benchmark how much longer builds are done now, maybe someone could find out?

But to be honest it was the first time I didn't had troubles setting different MySQL versions on travis. You can also add different vendors and it still works.

```yaml
# From:
  matrix:
    - MYSQL_IMAGE=mysql:5.5
    - MYSQL_IMAGE=mysql:5.6
    - MYSQL_IMAGE=mysql:5.7
```

```yaml
# To:
  matrix:
    - MYSQL_IMAGE=mysql:5.5
    - MYSQL_IMAGE=mysql:5.6
    - MYSQL_IMAGE=mysql:5.7
    - MYSQL_IMAGE=percona/percona-server:5.6
    - MYSQL_IMAGE=percona/percona-server:5.7
```

**Comment by nerrad** [comment]
Thanks for the workarounds @arvenil and @nazar-pc that you've posted here.  However, as @EnTeQuAk pointed out, part of the attraction of using Travis is being able to to builds on something similar to a production environment.  Hence using maria-db or a docker image container is not an ideal environment for the purpose of the tests we're running.  Ideally, it'd be nice to have this issue fixed in the default environment.

**Comment by nazar-pc** [comment]
I never meant this shouldn't be fixed, I'm with you on this :)

**Comment by BanzaiMan** [comment]
We have identified an issue with a kernel parameter that may have prevented MySQL instances from starting up. A change to it is now mostly in place (with a few exceptions).

Have you encountered this issue in the 24 hours or so?

**Comment by nerrad** [comment]
A job ran 4 hours ago (from this post) failed: https://travis-ci.org/eventespresso/event-espresso-core/jobs/257368318

**Comment by SuperTux88** [comment]
I had only 8 mysql jobs in the last 24h, but all were OK.

**Comment by nerrad** [comment]
It does look like more recent builds in various branches are more promising though (within the last hour).  We usually have a series of nightly builds happen across various repos overnight so I'll have a better idea tomorrow whether this is a solid improvement.

**Comment by BanzaiMan** [comment]
@nerrad @SuperTux88 Thanks for the reports. I can't tell if https://travis-ci.org/eventespresso/event-espresso-core/jobs/257368318 was run on a "bad" host (meaning one with the old value), but hopefully it was an exception rather than a rule.

**Comment by ntwb** [comment]
This job failed ~17 hours https://travis-ci.org/buddypress/BuddyPress/jobs/257217154 with the error:
> `ERROR 2002 (HY000): Can't connect to local MySQL server through socket '/var/run/mysqld/mysqld.sock' (2)`

_Will report back if I see any more jobs failing, I came here looking for more info as I've been seeing this for the past week or so on a different repo and slightly different (I think) to the fixed #8114 issue_

**Comment by nerrad** [comment]
Last night's nightlies all passed without this error (~20-30 jobs).  It's looking all good for us!  Is it something that just needed to replicate everywhere? Thanks.

I'll close this since it fixes the issue I reported.
